uberfire-extensions/uberfire-widgets/uberfire-widgets-commons/src/main/java/org/uberfire/ext/widgets/common/client/breadcrumbs/UberfireBreadcrumbs.java

```diff
@@ -255,86 +208,50 @@ public void addBreadCrumb(final String associatedPerspective,
      */
     public void addToolbar(final String associatedPerspective,
                            final Element toolbar) {
-        breadcrumbsToolBarPerPerspective.put(associatedPerspective,
-                                             toolbar);
-        if (currentPerspective == associatedPerspective) {
+
+        breadcrumbsToolBarPerPerspective.put(associatedPerspective, toolbar);
+        if (currentPerspective.equals(associatedPerspective)) {
             updateView();
         }
     }
 
-    private List<BreadcrumbsPresenter> getBreadcrumbs(final String perspective) {
-        List<BreadcrumbsPresenter> breadcrumbs = breadcrumbsPerPerspective.get(perspective);
-        if (breadcrumbs == null) {
-            breadcrumbs = new ArrayList<>();
-        }
-        return breadcrumbs;
+    private List<BreadcrumbPresenter> getBreadcrumbs(final String perspective) {
+        return breadcrumbsPerPerspective.getOrDefault(perspective, new ArrayList<>());
     }
 
-    private void deactivateLastBreadcrumb(final List<BreadcrumbsPresenter> breadcrumbs) {
+    private void activateLastBreadcrumb(final List<BreadcrumbPresenter> breadcrumbs) {
         if (!breadcrumbs.isEmpty()) {
-            breadcrumbs.get(breadcrumbs.size() - 1).deactivate();
+            breadcrumbs.get(breadcrumbs.size() - 1).activate();
         }
     }
 
-    private BreadcrumbsPresenter createBreadCrumb(final String perspective,
-                                                  final String label,
-                                                  final PlaceRequest placeRequest,
-                                                  final HasWidgets addTo,
-                                                  final Command command,
-                                                  final boolean removeDeepLevelBreadcrumbsAfterActivation) {
-
-        BreadcrumbsPresenter breadCrumb = breadcrumbsPresenters.get();
-        breadCrumb.setup(label,
-                         placeRequest,
-                         generateBreadCrumbSelectCommand(perspective,
-                                                         breadCrumb,
-                                                         placeRequest,
-                                                         addTo,
-                                                         command),
-                         removeDeepLevelBreadcrumbsAfterActivation);
-        breadCrumb.activate();
-        return breadCrumb;
-    }
-
     Command generateBreadCrumbSelectCommand(final String perspective,
-                                            final BreadcrumbsPresenter breadCrumb,
+                                            final DefaultBreadcrumbsPresenter breadCrumb,
                                             final PlaceRequest placeRequest,
-                                            final HasWidgets addTo,
                                             final Command command) {
+        if (command == null) {
```

Comment:
I see some usages of this method that directly call `execute()` on the result of this method. Wouldn't it be more safe to return "Command that does nothing" instead of null just to avoid potential future NPEs?
```java
if (command == null) {
    return () -> {};
}
```

Reply:
It was done using `null` here because the `DefaultBreadcrumbsPresenter.java` checks if the command is `null` in order to apply a specific style to it:
```
if (selectCommand == null) {
    view.setNoAction();
}
```

Before executing the command, the presenter also checks if it is not null:
```
public void onClick() {
    if (selectCommand != null) {
        selectCommand.execute();
    }
}
```